src/daft-micropartition/src/partitioning.rs

```diff
@@ -0,0 +1,208 @@
+    }
+    /// Number of partitions
+    fn num_partitions(&self) -> usize;
+
+    fn len(&self) -> usize;
+    /// Check if the partition set is empty
+    fn is_empty(&self) -> bool;
+    /// Size of the partition set in bytes
+    fn size_bytes(&self) -> DaftResult<usize>;
+    /// Check if a partition exists
+    fn has_partition(&self, idx: &PartId) -> bool;
+    /// Delete a partition
+    fn delete_partition(&mut self, idx: &PartId) -> DaftResult<()>;
+    /// Set a partition
+    fn set_partition(&mut self, idx: PartId, part: PartitionBatchRef) -> DaftResult<()>;
+    /// Get a partition
+    fn get_partition(&self, idx: &PartId) -> DaftResult<PartitionBatchRef>;
+}
+
+#[derive(Debug, Default)]
+pub struct InMemoryPartitionSet {
+    pub partitions: HashMap<String, Vec<Arc<MicroPartition>>>,
+}
+
+impl InMemoryPartitionSet {
+    pub fn new(psets: HashMap<String, Vec<Arc<MicroPartition>>>) -> Self {
+        Self { partitions: psets }
+    }
+}
+
+impl PartitionSet for InMemoryPartitionSet {
+    fn get_merged_micropartitions(&self) -> DaftResult<MicroPartition> {
+        let parts = self.values()?;
+        let parts = parts
+            .into_iter()
+            .flat_map(|mat_res| mat_res.micropartitions());
+
+        MicroPartition::concat(parts)
+    }
+
+    fn get_preview_micropartitions(
+        &self,
+        mut num_rows: usize,
+    ) -> DaftResult<Vec<Arc<MicroPartition>>> {
+        let mut preview_parts = vec![];
+
+        for part in self.partitions.values().flatten() {
+            let part_len = part.len();
+            if part_len >= num_rows {
+                preview_parts.push(Arc::new(part.slice(0, num_rows)?));
+                break;
+            } else {
+                num_rows -= part_len;
+                preview_parts.push(part.clone());
+            }
+        }
+        Ok(preview_parts)
+    }
+
+    fn items(&self) -> DaftResult<Vec<(PartId, PartitionBatchRef)>> {
+        self.partitions
+            .iter()
+            .map(|(k, v)| {
+                let partition = InMemoryPartitionBatch {
+                    partition: v.clone(),
+                    metadata: None,
+                };
+
+                Ok((k.clone(), Arc::new(partition) as _))
+            })
+            .collect()
+    }
+
+    fn num_partitions(&self) -> usize {
+        self.partitions.len()
+    }
+
+    fn len(&self) -> usize {
+        self.partitions.len()
+    }
+
+    fn is_empty(&self) -> bool {
+        self.partitions.is_empty()
+    }
+
+    fn size_bytes(&self) -> DaftResult<usize> {
+        let partitions = self.values()?;
+        let mut partitions = partitions.into_iter().flat_map(|mp| mp.micropartitions());
+        partitions.try_fold(0, |acc, mp| Ok(acc + mp.size_bytes()?.unwrap_or(0)))
+    }
+
+    fn has_partition(&self, partition_id: &PartId) -> bool {
+        self.partitions.contains_key(partition_id)
+    }
+
+    fn delete_partition(&mut self, partition_id: &PartId) -> DaftResult<()> {
```

Comment:
maybe add simple tests?

Reply:
right now there's quite a bit of setup for creating a micropartition in rust. I agree that we _should_ have tests, but our python implementation also didnt have any tests for this. So i think it should be fine for now.

Reply:
ok sgtm